pkg/apis/experimental/validation/validation.go

```diff
@@ -333,3 +345,130 @@ func ValidateJobSpecUpdate(oldSpec, spec experimental.JobSpec) errs.ValidationEr
 	}
 	return allErrs
 }
+
+// ValidateIngress tests if required fields in the Ingress are set.
+func ValidateIngress(ingress *experimental.Ingress) errs.ValidationErrorList {
+	allErrs := errs.ValidationErrorList{}
+	allErrs = append(allErrs, apivalidation.ValidateObjectMeta(&ingress.ObjectMeta, true, ValidateIngressName).Prefix("metadata")...)
+	allErrs = append(allErrs, ValidateIngressSpec(&ingress.Spec).Prefix("spec")...)
+
+	return allErrs
+}
+
+// ValidateIngressName validates that the given name can be used as an Ingress name.
+func ValidateIngressName(name string, prefix bool) (bool, string) {
+	return apivalidation.NameIsDNSSubdomain(name, prefix)
+}
+
+// ValidateIngressSpec tests if required fields in the IngressSpec are set.
+func ValidateIngressSpec(spec *experimental.IngressSpec) errs.ValidationErrorList {
+	allErrs := errs.ValidationErrorList{}
+	// TODO: Is a default backend mandatory?
+	if spec.Backend != nil {
+		allErrs = append(allErrs, validateIngressBackend(spec.Backend).Prefix("backend")...)
+	} else if len(spec.Hosts) == 0 {
+		allErrs = append(allErrs, errs.NewFieldInvalid("hosts", spec.Hosts, "Either a default backend or a set of host rules are required for ingress."))
+	}
+	if len(spec.Hosts) > 0 {
+		allErrs = append(allErrs, validateIngressHosts(spec.Hosts).Prefix("hosts")...)
+	}
+	return allErrs
+}
+
+// ValidateIngressUpdate tests if required fields in the Ingress are set.
+func ValidateIngressUpdate(oldController, ingress *experimental.Ingress) errs.ValidationErrorList {
+	allErrs := errs.ValidationErrorList{}
+	allErrs = append(allErrs, apivalidation.ValidateObjectMetaUpdate(&ingress.ObjectMeta, &oldController.ObjectMeta).Prefix("metadata")...)
+	allErrs = append(allErrs, ValidateIngressSpec(&ingress.Spec).Prefix("spec")...)
+	return allErrs
+}
+
+func validateIngressHosts(ingressHosts []experimental.IngressHost) errs.ValidationErrorList {
+	allErrs := errs.ValidationErrorList{}
+	if len(ingressHosts) == 0 {
+		return append(allErrs, errs.NewFieldRequired("ingressHosts"))
+	}
+	for _, ih := range ingressHosts {
+		if len(ih.Host) > 0 {
+			// TODO: Disallow ports?
+			host := strings.Split(ih.Host, ":")
+			if valid, errMsg := apivalidation.NameIsDNSSubdomain(host[0], false); !valid {
+				allErrs = append(allErrs, errs.NewFieldInvalid("host", host[0], errMsg))
+			}
+			if len(host) == 2 {
```

Comment:
port is allowed in host part of the url according to rfc 1738 (and the uri rfc, which is more recent) so I figured I'd just pass it through and let the loadbalancer controller decide what to do with it. I'm going for least restrictive validation that still produces something that roughly works cross-platform.